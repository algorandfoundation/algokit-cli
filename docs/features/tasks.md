# AlgoKit Tasks

AlgoKit Tasks are a collection of handy tasks that can be used to perform various operations on Algorand blockchain.

## Features

- [Wallet Aliasing](./tasks/wallet.md) - Manage your Algorand addresses and accounts effortlessly with the AlgoKit Wallet feature. This feature allows you to create short aliases for your addresses and accounts on AlgoKit CLI.
- [Vanity Address Generation](./tasks/vanity_address.md) - Generate vanity addresses for your Algorand accounts with the AlgoKit Vanity feature. This feature allows you to generate Algorand addresses which contains a specific keyword of your choice.
- [Transfer Assets or Algos](./tasks/transfer.md) - Transfer Algos or Assets from one account to another with the AlgoKit Transfer feature. This feature allows you to transfer Algos or Assets from one account to another on Algorand blockchain.
- [Opt-(in|out) Assets](./tasks/opt.md) - Opt-in or opt-out of Algorand Asset(s). Supports single or multiple assets.
- [Signing transactions](./tasks/sign.md) - Sign goal clerk compatible Algorand transactions.
- [Sending transactions](./tasks/send.md) - Send signed goal clerk compatible Algorand transactions.
- [NFD lookups](./tasks/nfd.md) - Perform a lookup via NFD domain or address, returning the associated address or domain respectively using the AlgoKit CLI.
- [IPFS uploads](./tasks/ipfs.md) - Upload files to IPFS.
- [Asset minting](./tasks/mint.md) - Mint new fungible or non-fungible assets on Algorand.
- [Analyze TEAL code](./tasks/analyze.md) - Analyze TEAL code using [`tealer`](https://github.com/crytic/tealer) integration for common vulnerabilities.
