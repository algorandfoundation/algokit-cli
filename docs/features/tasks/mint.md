# AlgoKit Task Mint

The AlgoKit Mint feature allows you to mint new fungible or non-fungible assets on the Algorand blockchain. This feature supports the creation of assets, validation of asset parameters, and uploading of asset metadata and image to IPFS using the Piñata provider. Immutable assets are compliant with [ARC3](https://arc.algorand.foundation/ARCs/arc-0003), while mutable are based using [ARC19](https://arc.algorand.foundation/ARCs/arc-0019) standard.

## Usage

Available commands and possible usage as follows:

```bash
Usage: algokit task mint [OPTIONS]

  Mint new fungible or non-fungible assets on Algorand.

Options:
  --creator TEXT                  Address or alias of the asset creator.  [required]
  -n, --name TEXT                 Asset name.  [required]
  -u, --unit TEXT                 Unit name of the asset.  [required]
  -t, --total INTEGER             Total supply of the asset. Defaults to 1.
  -d, --decimals INTEGER          Number of decimals. Defaults to 0.
  -i, --image FILE                Path to the asset image file to be uploaded to IPFS.  [required]
  -m, --metadata FILE             Path to the ARC19 compliant asset metadata file to be uploaded to IPFS. If not
                                  provided, a default metadata object will be generated automatically based on asset-
                                  name, decimals and image. For more details refer to
                                  https://arc.algorand.foundation/ARCs/arc-0003#json-metadata-file-schema.
  --mutable / --immutable         Whether the asset should be mutable or immutable. Refers to `ARC19` by default.
  --nft / --ft                    Whether the asset should be validated as NFT or FT. Refers to NFT by default and
                                  validates canonical definitions of pure or fractional NFTs as per ARC3 standard.
  -n, --network [localnet|testnet|mainnet]
                                  Network to use. Refers to `localnet` by default.
  -h, --help                      Show this message and exit.
```

## Options

- `--creator TEXT`: Specifies the address or alias of the asset creator. This option is required.
- `-n, --name TEXT`: Specifies the asset name. This option is required.
- `-u, --unit TEXT`: Specifies the unit name of the asset. This option is required.
- `-t, --total INTEGER`: Specifies the total supply of the asset. Defaults to 1.
- `-d, --decimals INTEGER`: Specifies the number of decimals. Defaults to 0.
- `-i, --image PATH`: Specifies the path to the asset image file to be uploaded to IPFS. This option is required.
- `-m, --metadata PATH`: Specifies the path to the ARC19 compliant asset metadata file to be uploaded to IPFS. If not provided, a default metadata object will be generated automatically based on asset-name, decimals and image.
- `--mutable / --immutable`: Specifies whether the asset should be mutable or immutable. Refers to `ARC19` by default.
- `--nft / --ft`: Specifies whether the asset should be validated as NFT or FT. Refers to NFT by default and validates canonical definitions of pure or fractional NFTs as per ARC3 standard.
- `-n, --network [localnet|testnet|mainnet]`: Specifies the network to use. Refers to `localnet` by default.

## Example

To mint a new asset in interactive mode, you can use the mint command as follows:

```bash
$ algokit task mint
```

This will interactively prompt you for the required information, upload the asset image and metadata to IPFS using the Piñata provider and mint a new asset on the Algorand blockchain. The [asset's metadata](https://arc.algorand.foundation/ARCs/arc-0003#json-metadata-file-schema) will be generated automatically based on the provided asset name, decimals, and image.

If you want to provide a custom metadata file, you can use the --metadata flag:

```bash
$ algokit task mint --metadata {PATH_TO_METADATA}
```

If the minting process is successful, the asset ID and transaction ID will be output to the console.

For non interactive mode, refer to usage section above for available options.

> Please note, creator account must have at least 0.2 Algos available to cover minimum balance requirements.

## Further Reading

For in-depth details, visit the [mint section](../../cli/index.md#mint) in the AlgoKit CLI reference documentation.
