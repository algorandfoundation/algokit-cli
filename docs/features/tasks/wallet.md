# AlgoKit Task Wallet

Manage your Algorand addresses and accounts effortlessly with the AlgoKit Wallet feature. This feature allows you to create short aliases for your addresses and accounts on AlgoKit CLI.

## Usage

Available commands and possible usage as follows:

```bash
$ ~ algokit task wallet
Usage: algokit task wallet [OPTIONS] COMMAND [ARGS]...

Create short aliases for your addresses and accounts on AlgoKit CLI.

Options:
-h, --help Show this message and exit.

Commands:
add Add an address or account to be stored against a named alias.
get Get an address or account stored against a named alias.
list List all addresses and accounts stored against a named alias.
remove Remove an address or account stored against a named alias.
reset Remove all aliases.
```

## Commands

### Add

This command adds an address or account to be stored against a named alias. If the `--mnemonic` flag is used, it will prompt the user for a mnemonic phrase interactively using masked input. If the `--force` flag is used, it will allow overwriting an existing alias. Maximum number of aliases that can be stored at a time is 50.

```bash
$ algokit wallet add [OPTIONS] ALIAS_NAME
```

> Please note, the command is not designed to be used in CI scope, there is no option to skip interactive masked input of the mnemonic, if you want to alias an `Account` (both private and public key) entity.

#### Options

- `--address, -a TEXT`: Specifies the address of the account. This option is required.
- `--mnemonic, -m`: If specified, it prompts the user for a mnemonic phrase interactively using masked input.
- `--force, -f`: If specified, it allows overwriting an existing alias without interactive confirmation prompt.

### Get

This command retrieves an address or account stored against a named alias.

```bash
$ algokit wallet get ALIAS
```

### List

This command lists all addresses and accounts stored against a named alias. If a record contains a `private_key` it will show a boolean flag indicating whether it exists, actual private key values are never exposed. As a user you can obtain the content of the stored aliases by navigating to your dedicated password manager (see [keyring details](https://pypi.org/project/keyring/)).

```bash
$ algokit wallet list
```

### Remove

This command removes an address or account stored against a named alias.
You must confirm the prompt interactively or pass `--force` | `-f` flag to ignore the prompt.

```bash
$ algokit wallet remove ALIAS  [--force | -f]
```

### Reset

This command removes all aliases. You must confirm the prompt interactively or pass `--force` | `-f` flag to ignore the prompt.

```bash
$ algokit wallet reset [--force | -f]
```

## Keyring

AlgoKit relies on the [keyring](https://pypi.org/project/keyring/) library, which provides an easy way to interact with the operating system's password manager. This abstraction allows AlgoKit to securely manage sensitive information such as mnemonics and private keys.

When you use AlgoKit to store a mnemonic, it is never printed or exposed directly in the console. Instead, the mnemonic is converted and stored as a private key in the password manager. This ensures that your sensitive information is kept secure.

To retrieve the stored mnemonic, you will need to manually navigate to your operating system's password manager. The keyring library supports a variety of password managers across different operating systems. Here are some examples:

- On macOS, it uses the Keychain Access app.
- On Windows, it uses the Credential Manager.
- On Linux, it can use Secret Service API, KWallet, or an in-memory store depending on your setup.

> Remember, AlgoKit is designed to keep your sensitive information secure however your storage is only as secure as the device on which it is stored. Always ensure to maintain good security practices on your device, especially when dealing with mnemonics that are to be used on MainNet.

## Further Reading

For in-depth details, visit the [wallet section](../../cli/index.md#wallet) in the AlgoKit CLI reference documentation.
