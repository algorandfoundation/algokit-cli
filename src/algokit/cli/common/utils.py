# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
