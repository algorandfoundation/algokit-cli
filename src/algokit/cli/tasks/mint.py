import logging
import math
from pathlib import Path

import click
from algosdk.error import AlgodHTTPError
from algosdk.util import algos_to_microalgos

from algokit.cli.common.constants import AlgorandNetwork, ExplorerEntityType
from algokit.cli.common.utils import get_explorer_url
from algokit.cli.tasks.utils import (
    get_account_with_private_key,
    load_algod_client,
    validate_balance,
)
from algokit.core.tasks.ipfs import (
    PinataBadRequestError,
    PinataForbiddenError,
    PinataHttpError,
    PinataInternalServerError,
    PinataUnauthorizedError,
    get_pinata_jwt,
)
from algokit.core.tasks.mint.mint import mint_token
from algokit.core.tasks.mint.models import TokenMetadata

logger = logging.getLogger(__name__)

MAX_UNIT_NAME_BYTE_LENGTH = 8
MAX_ASSET_NAME_BYTE_LENGTH = 32
ASSET_MINTING_MBR = 0.2  # Algos, 0.1 for base account, 0.1 for asset creation


def _validate_supply(total: int, decimals: int) -> None:
    """
    Validate the total supply and decimal places of a token.

    Args:
        total (int): The total supply of the token.
        decimals (int): The number of decimal places for the token.

    Raises:
        click.ClickException: If the validation fails.
    """
    if not (total == 1 or (total % 10 == 0 and total != 0)):
        raise click.ClickException("Total must be 1 or a power of 10 larger than 1 (10, 100, 1000, ...).")
    if not ((total == 1 and decimals == 0) or (total != 1 and decimals == int(math.log10(total)))):
        raise click.ClickException(
            "Number of digits after the decimal point must be 0 for a pure NFT, or "
            "equal to the logarithm in base 10 of total number of units for a fractional NFT."
        )


def _validate_unit_name(context: click.Context, param: click.Parameter, value: str) -> str:
    """
    Validate the unit name by checking if its byte length is less than or equal to a predefined maximum value.

    Args:
        context (click.Context): The click context.
        param (click.Parameter): The click parameter.
        value (str): The value of the parameter.

    Returns:
        str: The value of the parameter if it passes the validation.
    """

    if len(value.encode("utf-8")) <= MAX_UNIT_NAME_BYTE_LENGTH:
        return value
    else:
        raise click.BadParameter(
            f"Unit name must be {MAX_UNIT_NAME_BYTE_LENGTH} bytes or less.", ctx=context, param=param
        )


def _validate_asset_name(context: click.Context, param: click.Parameter, value: str) -> str:
    """
    Validate the asset name by checking if its byte length is less than or equal to a predefined maximum value.

    Args:
        context (click.Context): The click context.
        param (click.Parameter): The click parameter.
        value (str): The value of the parameter.

    Returns:
        str: The value of the parameter if it passes the validation.
    """

    if len(value.encode("utf-8")) <= MAX_ASSET_NAME_BYTE_LENGTH:
        return value
    else:
        raise click.BadParameter(
            f"Unit name must be {MAX_UNIT_NAME_BYTE_LENGTH} bytes or less.", ctx=context, param=param
        )


@click.command(
    name="mint",
    help="Mint new fungible or non-fungible assets on Algorand.",
)
@click.option(
    "--creator",
    required=True,
    prompt="Provide the address or alias of the asset creator",
    help="Address or alias of the asset creator.",
    type=click.STRING,
)
@click.option(
    "-n",
    "--name",
    "asset_name",
    type=click.STRING,
    required=True,
    callback=_validate_asset_name,
    prompt="Provide the asset name",
    help="Asset name.",
)
@click.option(
    "-u",
    "--unit",
    "unit_name",
    type=click.STRING,
    required=True,
    callback=_validate_unit_name,
    prompt="Provide the unit name",
    help="Unit name of the asset.",
)
@click.option(
    "-t",
    "--total",
    type=click.INT,
    required=False,
    default=1,
    prompt="Provide the total supply",
    help="Total supply of the asset. Defaults to 1.",
)
@click.option(
    "-d",
    "--decimals",
    type=click.INT,
    required=False,
    default=0,
    prompt="Provide the number of decimals",
    help="Number of decimals. Defaults to 0.",
)
@click.option(
    "-i",
    "--image",
    "image_path",
    type=click.Path(exists=True, dir_okay=False, file_okay=True, resolve_path=True, path_type=Path),
    prompt="Provide the path to the asset image file",
    help="Path to the asset image file to be uploaded to IPFS.",
    required=True,
)
@click.option(
    "-m",
    "--metadata",
    "token_metadata_path",
    type=click.Path(exists=True, dir_okay=False, file_okay=True, resolve_path=True, path_type=Path),
    help="""Path to the ARC19 compliant asset metadata file to be uploaded to IPFS. If not provided,
        a default metadata object will be generated automatically based on asset-name, decimals and image.
        For more details refer to https://arc.algorand.foundation/ARCs/arc-0003#json-metadata-file-schema.""",
    default=None,
    required=False,
)
@click.option(
    "--mutable/--immutable",
    type=click.BOOL,
    prompt="Would you like to make the asset mutable?",
    default=False,
    help="Whether the asset should be mutable or immutable. Refers to `ARC19` by default.",
)
@click.option(
    "--nft/--ft",
    "non_fungible",
    type=click.BOOL,
    prompt="Validate asset as NFT? Checks values of `total` and `decimals` as per ARC3 if set to True.",
    default=False,
    help="""Whether the asset should be validated as NFT or FT. Refers to NFT by default and validates canonical
    definitions of pure or fractional NFTs as per ARC3 standard.""",
)
@click.option(
    "-n",
    "--network",
    type=click.Choice(AlgorandNetwork.to_list()),
    prompt="Provide the network to use",
    default=AlgorandNetwork.LOCALNET,
    required=False,
    help=f"Network to use. Refers to `{AlgorandNetwork.LOCALNET}` by default.",
)
def mint(  # noqa: PLR0913
    *,
    creator: str,
    asset_name: str,
    unit_name: str,
    total: int,
    decimals: int,
    image_path: Path,
    token_metadata_path: Path | None,
    mutable: bool,
    non_fungible: bool,
    network: AlgorandNetwork,
) -> None:
    if non_fungible:
        _validate_supply(total, decimals)

    creator_account = get_account_with_private_key(creator)

    pinata_jwt = get_pinata_jwt()
    if not pinata_jwt:
        raise click.ClickException("You are not logged in! Please login using `algokit task ipfs login`.")

    client = load_algod_client(network)
    validate_balance(
        client,
        creator_account,
        0,
        algos_to_microalgos(ASSET_MINTING_MBR),  # type: ignore[no-untyped-call]
    )

    token_metadata = TokenMetadata.from_json_file(token_metadata_path)
    if not token_metadata_path:
        token_metadata.name = asset_name
        token_metadata.decimals = decimals
    try:
        asset_id, txn_id = mint_token(
            client=client,
            jwt=pinata_jwt,
            creator_account=creator_account,
            token_metadata=token_metadata,
            image_path=image_path,
            unit_name=unit_name,
            asset_name=asset_name,
            mutable=mutable,
            total=total,
        )

        click.echo("\nSuccessfully minted the asset!")
        click.echo(f"Browse your asset at: {get_explorer_url(asset_id, network, ExplorerEntityType.ASSET)}")
        click.echo(f"Check transaction status at: {get_explorer_url(txn_id, network, ExplorerEntityType.TRANSACTION)}")
    except (
        PinataBadRequestError,
        PinataUnauthorizedError,
        PinataForbiddenError,
        PinataInternalServerError,
        PinataHttpError,
    ) as ex:
        logger.debug(ex)
        raise click.ClickException(repr(ex)) from ex
    except AlgodHTTPError as ex:
        raise click.ClickException(str(ex)) from ex
    except Exception as ex:
        logger.debug(ex, exc_info=True)
        raise click.ClickException("Failed to mint the asset!") from ex
