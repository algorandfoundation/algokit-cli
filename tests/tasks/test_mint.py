import json
import re
from pathlib import Path

import click
import pytest
from algosdk.mnemonic import from_private_key
from approvaltests.namer import NamerFactory
from pytest_httpx import HTTPXMock
from pytest_mock import MockerFixture

from algokit.cli.tasks.mint import _get_and_validate_asset_name, _get_and_validate_decimals
from algokit.core.tasks.wallet import WALLET_ALIASES_KEYRING_USERNAME
from tests.tasks.conftest import DUMMY_ACCOUNT, DUMMY_SUGGESTED_PARAMS
from tests.utils.approvals import verify
from tests.utils.click_invoker import invoke


@pytest.mark.parametrize(("account_type", "mutability"), [("alias", "mutable"), ("address", "immutable")])
@pytest.mark.parametrize("network", ["localnet", "testnet", "mainnet"])
def test_mint_token_successful(
    *,
    mocker: MockerFixture,
    tmp_path_factory: pytest.TempPathFactory,
    mock_keyring: dict[str, str | int],
    account_type: str,
    mutability: str,
    network: str,
) -> None:
    # Arrange
    is_mutable = mutability == "mutable"
    cwd = tmp_path_factory.mktemp("cwd")
    account = ""
    prompt_input = None
    if account_type == "address":
        account = DUMMY_ACCOUNT.address
        prompt_input = from_private_key(DUMMY_ACCOUNT.private_key)  # type: ignore[no-untyped-call]
    else:
        account = "my_alias"
        mock_keyring[account] = json.dumps(
            {
                "alias": account,
                "address": DUMMY_ACCOUNT.address,
                "private_key": DUMMY_ACCOUNT.private_key,
            }
        )
        mock_keyring[WALLET_ALIASES_KEYRING_USERNAME] = json.dumps([account])
    (cwd / "image.png").touch()

    mocker.patch(
        "algokit.core.tasks.mint.mint.upload_to_pinata",
        side_effect=[
            "bafkreifax6dswcxk4us2am3jxhd3swxew32oreaxzol7dnnqzhieepqg2y",
            "bafkreiftmc4on252dnckhv7jdqnhkxjkpvlrekpevjwm3gjszygxkus5oe",
        ],
    )
    mocker.patch("algokit.core.tasks.mint.mint.wait_for_confirmation", return_value={"asset-index": 123})
    mocker.patch(
        "algokit.cli.tasks.mint.get_pinata_jwt",
        return_value="dummy_key",
    )
    mocker.patch(
        "algokit.cli.tasks.mint.validate_balance",
    )
    algod_mock = mocker.MagicMock()
    algod_mock.send_transaction.return_value = "dummy_tx_id"
    algod_mock.suggested_params.return_value = DUMMY_SUGGESTED_PARAMS
    mocker.patch("algokit.cli.tasks.mint.load_algod_client", return_value=algod_mock)

    # Act
    result = invoke(
        f"""task mint --creator {account} --name test --unit tst --total 1 --decimals 0
        --image image.png -n {network} --{"mutable" if is_mutable else "immutable"} --nft""",
        input=prompt_input,
        cwd=cwd,
    )

    # Assert
    assert result.exit_code == 0
    if is_mutable:
        # Reserve value must be set since its a mutable asset
        assert re.search(r'"reserve": ".{58}"', result.output) is not None, (
            "Reserve key not found or addr length is not 58"
        )
    else:
        assert re.search(r'"reserve": ""', result.output) is not None, "Reserve key must be empty"
    verify(result.output, options=NamerFactory.with_parameters(account_type, is_mutable, network))


@pytest.mark.parametrize("decimals", ["decimals_given_params", "no_decimals_given"])
def test_mint_token_successful_on_decimals(
    *,
    mocker: MockerFixture,
    tmp_path_factory: pytest.TempPathFactory,
    mock_keyring: dict[str, str | int],
    decimals: str,
) -> None:
    # Arrange
    cwd = tmp_path_factory.mktemp("cwd")
    if decimals == "no_decimals_given":
        include_decimals_argument = False
        prompt_input = "2"
    elif decimals == "decimals_given_params":
        include_decimals_argument = True
        prompt_input = None

    account = "my_alias"
    mock_keyring[account] = json.dumps(
        {
            "alias": account,
            "address": DUMMY_ACCOUNT.address,
            "private_key": DUMMY_ACCOUNT.private_key,
        }
    )
    mock_keyring[WALLET_ALIASES_KEYRING_USERNAME] = json.dumps([account])

    (cwd / "image.png").touch()

    mocker.patch(
        "algokit.core.tasks.mint.mint.upload_to_pinata",
        side_effect=[
            "bafkreifax6dswcxk4us2am3jxhd3swxew32oreaxzol7dnnqzhieepqg2y",
            "bafkreiftmc4on252dnckhv7jdqnhkxjkpvlrekpevjwm3gjszygxkus5oe",
        ],
    )
    mocker.patch("algokit.core.tasks.mint.mint.wait_for_confirmation", return_value={"asset-index": 123})
    mocker.patch(
        "algokit.cli.tasks.mint.get_pinata_jwt",
        return_value="dummy_key",
    )
    mocker.patch(
        "algokit.cli.tasks.mint.validate_balance",
    )
    algod_mock = mocker.MagicMock()
    algod_mock.send_transaction.return_value = "dummy_tx_id"
    algod_mock.suggested_params.return_value = DUMMY_SUGGESTED_PARAMS
    mocker.patch("algokit.cli.tasks.mint.load_algod_client", return_value=algod_mock)

    # Act
    result = invoke(
        f"""task mint --creator {account} --name test --unit tst --total 100
        {"--decimals 2 " if include_decimals_argument else ""}
        --image image.png -n localnet --mutable --nft""",
        input=prompt_input,
        cwd=cwd,
    )

    # Assert
    assert result.exit_code == 0
    verify(result.output, options=NamerFactory.with_parameters(decimals))


def test_mint_token_pure_fractional_nft_ft_validation(
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    # Arrange
    network = "localnet"
    cwd = tmp_path_factory.mktemp("cwd")
    account = DUMMY_ACCOUNT.address
    prompt_input = from_private_key(DUMMY_ACCOUNT.private_key)  # type: ignore[no-untyped-call]
    (cwd / "image.png").touch()

    # Act
    nft_result = invoke(
        f"""task mint --creator {account} --name test --unit tst --total 222 --decimals 12
        --image image.png -n {network} --mutable --nft""",
        input=prompt_input,
        cwd=cwd,
    )

    # Assert
    assert nft_result.exit_code == 1


def test_mint_token_pinata_error(
    mocker: MockerFixture,
    httpx_mock: HTTPXMock,
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    # Arrange
    cwd = tmp_path_factory.mktemp("cwd")
    account = ""
    account = DUMMY_ACCOUNT.address
    prompt_input = from_private_key(DUMMY_ACCOUNT.private_key)  # type: ignore[no-untyped-call]
    (cwd / "image.png").touch()
    httpx_mock.add_response(status_code=403, json={"ok": False})

    mocker.patch(
        "algokit.cli.tasks.mint.get_pinata_jwt",
        return_value="dummy_key",
    )
    mocker.patch(
        "algokit.cli.tasks.mint.validate_balance",
    )
    algod_mock = mocker.MagicMock()
    mocker.patch("algokit.cli.tasks.mint.load_algod_client", return_value=algod_mock)

    # Act
    result = invoke(
        f"""task mint --creator {account} --name test --unit tst --total 1 --decimals 0
        --image image.png -n localnet --mutable --nft""",
        input=prompt_input,
        cwd=cwd,
    )

    # Assert
    assert result.exit_code == 1
    verify(result.output)


def test_mint_token_no_pinata_jwt_error(
    mocker: MockerFixture,
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    # Arrange
    cwd = tmp_path_factory.mktemp("cwd")
    account = ""
    account = DUMMY_ACCOUNT.address
    prompt_input = from_private_key(DUMMY_ACCOUNT.private_key)  # type: ignore[no-untyped-call]
    (cwd / "image.png").touch()

    mocker.patch(
        "algokit.cli.tasks.mint.get_pinata_jwt",
        return_value=None,
    )
    # Act
    result = invoke(
        f"""task mint --creator {account} --name test --unit tst --total 1 --decimals 0
        --image image.png -n localnet --mutable --nft""",
        cwd=cwd,
        input=prompt_input,
    )

    # Assert
    assert result.exit_code == 1
    verify(result.output)


def test_mint_token_acfg_token_metadata_mismatch_on_name(
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / "metadata.json").write_text(
        """
        {
        "name": "test2",
        "decimals": 2,
        "description": "Stars",
        "properties": {
            "author": "Al",
            "traits": {
            "position": "center",
            "colors": 4
            }
        }
        }
        """
    )
    context = click.Context(click.Command("mint"))
    context.params["token_metadata_path"] = Path(cwd / "metadata.json")
    param = click.Option(["--name"])
    name = "test"

    with pytest.raises(
        click.BadParameter, match="Token name in metadata JSON must match CLI argument providing token name!"
    ):
        _get_and_validate_asset_name(context, param, name)


def test_mint_token_acfg_token_metadata_mismatch_on_decimals(
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / "metadata.json").write_text(
        """
        {
        "name": "test2",
        "decimals": 2,
        "description": "Stars",
        "properties": {
            "author": "Al",
            "traits": {
            "position": "center",
            "colors": 4
            }
        }
        }
        """
    )
    context = click.Context(click.Command("mint"))
    context.params["token_metadata_path"] = Path(cwd / "metadata.json")
    param = click.Option(["--decimals"])
    decimals = 0

    with pytest.raises(
        click.BadParameter, match="The value for decimals in the metadata JSON must match the decimals argument"
    ):
        _get_and_validate_decimals(context, param, decimals)
