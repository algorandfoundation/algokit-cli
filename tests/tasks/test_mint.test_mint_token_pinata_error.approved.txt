Enter the mnemonic phrase (25 words separated by whitespace): 
Uploading image to pinata...
HTTP Request: POST https://api.pinata.cloud/pinning/pinFileToIPFS "HTTP/1.1 403 Forbidden"
DEBUG: Pinata error: 403. {"ok":false}
Error: PinataForbiddenError('Pinata error: 403')
